src/ngraph/runtime/opt_kernel/broadcast.hpp

```diff
@@ -0,0 +1,235 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <cmath>
+#include <utility>
+
+#include "ngraph/runtime/reference/broadcast.hpp"
+#include "ngraph/shape_util.hpp"
+#include "ngraph/util.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace opt_kernel
+        {
+            template <typename T>
+            void broadcast_2d(const T* in,
+                              T* out,
+                              const Shape& in_shape,
+                              const Shape& out_shape,
+                              const AxisSet& broadcast_axes)
+            {
+                size_t index[2];
+                size_t* out_index =
+                    (broadcast_axes.find(0) == broadcast_axes.end() ? &index[0] : &index[1]);
+                for (index[0] = 0; index[0] < out_shape[0]; ++index[0])
```

Comment:
For the index offset calculations here and below, use `row_major_strides` to get the strides once outside the loop and then do a simple dot product of the index and strides.